driver-core/src/main/com/mongodb/internal/authentication/AwsInternalCredential.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.authentication;
+
+import com.mongodb.AwsCredential;
+import com.mongodb.MongoInternalException;
+import com.mongodb.lang.NonNull;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonDocument;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+
+public final class AwsInternalCredential {
+    private final String accessKeyId;
+    private final String secretAccessKey;
+    private final String sessionToken;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param accessKeyId     the non-null access key ID that identifies the temporary security credentials.
+     * @param secretAccessKey the non-null secret access key that can be used to sign requests
+     * @param sessionToken    the non-null session token
+     */
+    public AwsInternalCredential(@Nullable final String accessKeyId, @Nullable final String secretAccessKey,
+            @Nullable final String sessionToken) {
+        this.accessKeyId = accessKeyId;
+        this.secretAccessKey = secretAccessKey;
+        this.sessionToken = sessionToken;
+    }
+
+    /**
+     * Gets the access key ID that identifies the temporary security credentials.
+     *
+     * @return the accessKeyId, which may not be null
+     */
+    @Nullable
+    public String getAccessKeyId() {
+        return accessKeyId;
+    }
+
+    /**
+     * Gets the secret access key that can be used to sign requests.
+     *
+     * @return the secretAccessKey, which may not be null
+     */
+    @Nullable
+    public String getSecretAccessKey() {
+        return secretAccessKey;
+    }
+
+    /**
+     * Gets the session token.
+     *
+     * @return the sessionToken, which may not be null
+     */
+    @Nullable
+    public String getSessionToken() {
+        return sessionToken;
+    }
+
+    public static AwsInternalCredential obtainFromEnvironment(final AwsInternalCredential startingCredential,
```

Comment:
I opted for this static factory method.  It's a bit strange because it both takes and returns an instance of itself.  

Happy to discuss alternative designs.